### PR TITLE
Added test/correctness/conv3x3a16.cpp

### PR DESCRIPTION
**Pull request by dpalermo**
test/correctness/conv3x3a16.cpp

•	no performance measurement
•	synthetic input
•	checked internally by test
•	no external images used

### CODE REVIEW

File: test/correctness/conv3x3a16.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    if (verbose) {
+        for (int y = 0; y < 3; y++) {
+            for (int x = 0; x < 3; x++) {
+                printf("mask(%d, %d) = %d\n", x, y, mask(x, y));
+            }
+        }
+    }
+
+// Boundary condition
+#define BIN(x, y) (((x)>=0 && (x)<W && (y)>=0 && (y)< H) ? in((x), (y)) : exterior)
```
Comment (abadams) [inline test/correctness/conv3x3a16.cpp]: This might be better as a lambda than a macro:

    auto bin = [&](int x, int y) { return (x >= 0 && x < W && y >=0 && y < H) ? in(x, y) : exterior; };

Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: Good idea.  I've changed bounded input macro into a lambda.

File: test/correctness/conv3x3a16.cpp [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+#ifdef RDOM_OK
+    conv3x3(x, y) =
+        cast<uint8_t>(clamp(
+            sum(cast<int16_t>(input(x+r.x, y+r.y)) *
+                cast<int16_t>(mask(1+r.x, 1+r.y))) >> 4,
+            0, 255));
+#else
+    conv3x3(x, y) =
+        cast<uint8_t>(clamp(
+            (cast<int32_t>(input(x-1, y-1)) * cast<int32_t>(mask(0,0)) +
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: Note that the RDOM_OK code widens to int16_t, this widens to int32_t. The generated could would be quite different: The RDom version would generate 8 -> 16 bit widening multiply adds, the not-RDom code would widen to 16 bits, and then generate 16 -> 32 bit widening multiply adds. I'd expect the non-RDom path to be *at least* 2x slower than it should be (half the number of ops/cycle, plus the cost of casting).

Also note that the validation below uses int16, so if the mask coefficients were large enough to cause overflow of a 16 bit accumulator, I think this test would fail :)


Also, something I do in code like this is just use a compile time loop to generate the expression:

    Expr sum = 0;
    for (int i = -1; i <= 1; i++) {
        for (int j = -1; j <= 1; j++) {
            sum += cast<int32_t>(input(x + j, y + i)) * cast<int32_t>(mask(j + 1, i + 1));
        }
    }
    conv3x3(x, y) = cast<uint8_t>(clamp(sum >> 4, 0, 255));

File: test/correctness/conv3x3a16.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+#endif
+
+    // Schedule.
+    Target target = get_jit_target_from_environment();
+    if (target.has_gpu_feature()) {
+        Var xi("xi"), yi("yi");
+        conv3x3.gpu_tile(x, y, xi, yi, 16, 16);
+    } else if (target.features_any_of({Target::HVX_64, Target::HVX_128})) {
+        int vector_size = target.has_feature(Target::HVX_128) ? 128 : 64;
+        printf("HVX vector_size = %d\n", vector_size);
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: I think you should get rid of this printf.

File: test/correctness/conv3x3a16.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        Var xi("xi"), yi("yi");
+        conv3x3.gpu_tile(x, y, xi, yi, 16, 16);
+    } else if (target.features_any_of({Target::HVX_64, Target::HVX_128})) {
+        int vector_size = target.has_feature(Target::HVX_128) ? 128 : 64;
+        printf("HVX vector_size = %d\n", vector_size); 
+        conv3x3.hexagon().vectorize(x, vector_size);
+
+        conv3x3.output_buffer().dim(0).set_min(0);
+        conv3x3.output_buffer().dim(1).set_min(0);
+        conv3x3.output_buffer().set_host_alignment(vector_size);
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: set_host_alignment is unnecessary when using the hexagon() directive (that assumes the Hexagon buffers are always aligned, because either it allocated the memory itself and so it can assume it is aligned, or, the FastRPC interface is going to allocate the buffer and copy to it, which we also assume is aligned).

File: test/correctness/conv3x3a16.cpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    } else {
+        conv3x3.vectorize(x, target.natural_vector_size<uint8_t>());
+    }
+
+    // Run the pipeline and verify the results are correct.
+    Buffer<uint8_t> out = conv3x3.realize(W, H, target);
+
+
+    if (verbose) {
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: @abadams @srj @zvookin what are your thoughts on including code like this in the correctness tests? (note `int verbose = 0` at the top). I've added similar code to test/correctness/convolution.cpp many times while debugging something using that test, but I've never committed those changes, I'm leaning towards suggesting we continue to not check in code like that here too...

File: test/correctness/conv3x3a16.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
+#include "Halide.h"
+#include <iostream>
+
+//#define RDOM_OK 1
+
+using namespace Halide;
+using namespace Halide::Internal;
+
+int main(int arch, char **argv) {
+    int verbose = 0;
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: If we keep this (see comment below on the usage of this), this probably should be const bool.

File: test/correctness/conv3x3a16.cpp (hunk 2 above)
Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: Sorry, those should have all been int16_t (conv3x3a16) not int32_t (conv3x3a32)...fixed.  For conv3x3a16, the mask coefficients are assumed to not overflow.  The 32-bit accumulator version (conv3x3a32) is for larger mask values.

The compile time loop is pretty cool.  I've made that change as well (with int16_t)

File: test/correctness/conv3x3a16.cpp (hunk 3 above)
Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: done

File: test/correctness/conv3x3a16.cpp (hunk 4 above)
Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: removed

File: test/correctness/conv3x3a16.cpp (hunk 5 above)
Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: I used this to debug the test, I can remove it if you like.  I tend to like keeping useful debug code around but disabled in case it is ever needed in the future (also leaves nice hints as to what was useful to look at when debugging the code the last time).

File: test/correctness/conv3x3a16.cpp (hunk 2 above)
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: I got a bit carried away there, I was hoping that the int32 vs. int16 thing explained some of the performance gap we're seeing... :)

File: test/correctness/conv3x3a16.cpp (hunk 5 above)
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: My vote would be to remove it. Like I said I've needed it for other tests too, but since no other tests have debugging code like this in them already (that I've seen), I just added the debugging code I needed locally and removed it prior to committing my changes.

Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: No problem.  I removed it.

File: test/correctness/conv3x3a16.cpp (hunk 2 above)
Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: Sorry about that, that would have been nice ;-).  The version we are measuring for performance testing is indeed using int16_t.   This came in here, as I was actually chasing a down a miscompare bug when I first put this version together and bumped up the types to make sure it wasn't overflow...turned out to be a missing clamp in the C checker....and I forgot to bump the types back down.

File: test/correctness/conv3x3a16.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        for (int x = 0; x < W; x++) {
+            int16_t correct = (mask(0,0)*bin(x-1, y-1) + mask(1,0)*bin(x, y-1) + mask(2,0)*bin(x+1, y-1) +
+                               mask(0,1)*bin(x-1, y)   + mask(1,1)*bin(x, y)   + mask(2,1)*bin(x+1, y)   +
+                               mask(0,2)*bin(x-1, y+1) + mask(1,2)*bin(x, y+1) + mask(2,2)*bin(x+1, y+1)) >> 4;
+            // clamp
+            if (correct > 255) { correct = 255; }
+            if (correct < 0) { correct = 0; }
+
+            if (correct != out(x, y)) {
+                // printf("in(%d, %d) = %d\n", x, y, in(x, y));
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: Should remove this commented code.

Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: Done

**Comment by ronlieb** [comment]
LGTM, ok to merge now ?

**Comment by dsharletg** [comment]
After talking offline, it sounds like the plan is to write these tests in a way that can be tested on device, not only on the simulator. So, we're going to make an app that has all of these performance tests in it. If that's the case, I think we shouldn't merge this, as the same thing will be in the test app.

File: test/correctness/conv3x3a16.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    RDom r(-1,3,-1,3);
+    Func conv3x3("conv3x3a16");
+#ifdef RDOM_OK
+    conv3x3(x, y) =
+        cast<uint8_t>(clamp(
+            sum(cast<int16_t>(input(x+r.x, y+r.y)) *
+                cast<int16_t>(mask(1+r.x, 1+r.y))) >> 4,
+            0, 255));
+#else
+    Expr sum = 0;
```
Comment (dsharletg) [inline test/correctness/conv3x3a16.cpp]: Just noticed this needs to be cast<int16_t>(0), otherwise the reduction is done in 32 bits.

Comment (dpalermo) [inline test/correctness/conv3x3a16.cpp]: Fixed

**Comment by dsharletg** [comment]
Closing this because we are about to merge #1887.
